File: HW4/hw4/gradientMag.m
function gradientMag (im,thresh)         % viz grad magnitude of grayscale image

  % im: float over [0,1]: use `im=double(imread(fn))/255.;`
  % thresh: min gradient magnitude for bitmap (e.g., .5)

  if strcmp(class(im), 'uint8') || (max(im(:)) > 1)
      fprintf('please, I want a float [0,1] image\n')
      return
      end
      
  [gx gy] = imgradientxy (im, 'sobel');     % Cartesian components of gradient
                                            % see other options in documentation
  [gmag gdir] = imgradient (gx, gy);        % polar components of gradient
  fprintf("max gmag: %f\n", max(gmag(:)));
  fprintf("mean gmag: %f\n", mean(gmag(:)));
  gmag_thresh = gmag >= thresh;
    	      
  
  subplot (2,2,1); imshow (im);                         % over [0,1]
                   title ('I');
  subplot (2,2,2); imshow ((gx+4)/8);                   % over [-4,4] 
		   title ('df/dx');                     % why [-4,4] if Sobel?
		                                        % alt: imshow(gx,[-4,4])
  subplot (2,2,3); imshow (gmag / (4*sqrt(2)));         
                   title ('gradient magnitude');        
  subplot (2,2,4); imshow (gmag_thresh);                % over [0,1] (a bitmap)
                   title ('thresholded gmag');
  

File: HW4/hw4/gradientMag.py
import cv2 as cv
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.image as im
from statistics import mean as mean
import numpy as np
from math import sqrt


def gradientMag(image, thresh):
    """Filter the edges of an image by the angle "direction"
    Params: image (ndarray): An image
    thresh (int): Minimum gradient mag for bitmap
    Displays: Pyplot of four gmag visuals
    """

    gmag = image

    gaussian = cv.GaussianBlur(image, (3, 3), 0)
    print("Max gmag", np.max(gmag))
    print("Mean gmag", np.mean(gmag))
    s_horizontal = cv.Sobel(image, cv.CV_8U, 1, 0, ksize=5)
    s_vertical = cv.Sobel(image, cv.CV_8U, 0, 1, ksize=5)

    gmag_thresh = gmag <= thresh

    fig, axs = plt.subplots(2, 2)
    fig.suptitle("")

    axs[0, 0].imshow(gmag, cmap="gray")

    axs[0, 1].imshow(((s_horizontal + 4) / 8), cmap="gray")

    axs[1, 0].imshow(s_vertical / (4 * sqrt(2)), cmap="gray")

    axs[1, 1].imshow(gmag_thresh, cmap="gray")

    plt.show()


if __name__ == "__main__":
    image = im.imread("graypeppers.jpg")
    gradientMag(image, 1)
